File: daily/2024/2779_Maximum_Beauty_of_an_Array_After_Applying_Operation.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int maximumBeauty(vector<int>& nums, int k) {
        int maxv = 0;
        for (int i : nums) maxv = max(i, maxv);
        vector<int> count(maxv+1, 0);
        for (int i : nums){
            count[max(0, i-k)]++;
            if (i+k+1 <= maxv) count[i+k+1]--;
        }
        int ans = 0;
        int curr = 0;
        for (int i : count){
            curr += i;
            ans = max(ans, curr);
        }
        return ans;
    }
};

int main(){
    vector<int> nums = {4,6,1,2};
    int k = 2;
    // vector<int> nums = {1,1,1,1};
    // int k = 10;
    Solution S;
    cout << S.maximumBeauty(nums, k) << endl;
    return 0;
}

File: daily_python/2779_Maximum_Beauty_of_an_Array_After_Applying_Operation.py
import os
import sys
from typing import List

class Solution:
    def maximumBeauty(self, nums: List[int], k: int) -> int:
        maxv = max(nums)
        count = [0] * (maxv+1)
        for num in nums:
            count[max(0, num-k)] += 1
            if num+k+1 <= maxv:
                count[num+k+1] -= 1
        ans = 0
        curr = 0
        for i in count:
            curr += i
            ans = max(ans, curr)
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            k = int(lines[1].split('=')[1].strip()[:-1])
            dataset.append((nums, k))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums, k in dataset:
        results.append(solution.maximumBeauty(nums, k))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')